// ng-packagr/lib/flatten/uglify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.minifyJsFile = void 0;
const terser_1 = require("terser");
const path_1 = require("path");
const log_1 = require("../utils/log");
const fs_1 = require("../utils/fs");
async function minifyJsFile(inputPath, outputPath) {
    log_1.debug(`minifyJsFile: ${inputPath}`);
    const sourcemapOut = `${outputPath}.map`;
    const [inputFileBuffer, inputSourceMapBuffer] = await Promise.all([
        fs_1.readFile(inputPath),
        fs_1.readFile(`${inputPath}.map`),
    ]);
    const result = await terser_1.minify(inputFileBuffer.toString(), {
        sourceMap: {
            includeSources: true,
            content: JSON.parse(inputSourceMapBuffer.toString()),
            url: path_1.basename(sourcemapOut),
        },
        parse: {
            ecma: 5,
            bare_returns: true,
        },
        format: {
            comments: 'some',
        },
    });
    await Promise.all([fs_1.writeFile(outputPath, result.code), fs_1.writeFile(sourcemapOut, result.map)]);
    return;
}
exports.minifyJsFile = minifyJsFile;
