// ng-packagr/lib/utils/fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.copyFile = exports.exists = exports.stat = exports.mkdir = exports.access = exports.writeFile = exports.readFile = exports.rimraf = void 0;
const util_1 = require("util");
const rm = require("rimraf");
const fs = require("fs");
const path_1 = require("path");
exports.rimraf = util_1.promisify(rm);
exports.readFile = util_1.promisify(fs.readFile);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.access = util_1.promisify(fs.access);
exports.mkdir = util_1.promisify(fs.mkdir);
exports.stat = util_1.promisify(fs.stat);
async function exists(path) {
    try {
        await exports.access(path, fs.constants.F_OK);
        return true;
    }
    catch (_a) {
        return false;
    }
}
exports.exists = exists;
const cpFile = util_1.promisify(fs.copyFile);
async function copyFile(src, dest) {
    const dir = path_1.dirname(dest);
    if (!(await exists(dir))) {
        await exports.mkdir(dir, { recursive: true });
    }
    await cpFile(src, dest, fs.constants.COPYFILE_FICLONE);
}
exports.copyFile = copyFile;
